airflow/executors/celery_executor.py

```diff
@@ -274,6 +331,61 @@ def execute_async(self,
     def terminate(self):
         pass
 
+    def try_adopt_task_instances(self, tis: List[TaskInstance]) -> List[TaskInstance]:
+        # See which of the TIs are still alive (or have finished even!)
+        #
+        # Since Celery doesn't store "SENT" state for queued commands (if we create an AsyncResult with a made
+        # up id it just returns PENDING state for it), we have to store Celery's task_id against the TI row to
+        # look at in future.
+        #
+        # This process is not perfect -- we could have sent the task to celery, and crashed before we were
+        # able to record the AsyncResult.task_id in the TaskInstance table, in which case we won't adopt the
+        # task (it'll either run and update the TI state, or the scheduler will clear and re-queue it. Either
+        # way it won't get executed more than once)
+        #
+        # (If we swapped it around, and generated a task_id for Celery, stored that in TI and enqueued that
+        # there is also still a race condition where we could generate and store the task_id, but die before
+        # we managed to enqueue the command. Since neither way is perfect we always have to deal with this
+        # process not being perfect.)
+
+        celery_tasks = {
+            ti.external_executor_id: (AsyncResult(ti.external_executor_id), ti)
+            for ti in tis
+            if ti.external_executor_id is not None
+        }
```

Comment:
Not a big thing but we can do one `for` loop instead of this list comprehension and list in L365, right?

Reply:
We could, but they have different conditions (`if` and `if not`) - we'd need to do something like:

```python
celery_tasks = []
not_adopted = []

for ti in its:
  if ti.external_executor_id:
    celery_tasks.append(...)
  else:
    not_adopted.append(...)
```

Is that what you meant?

Reply:
Yes, I don't think this changes performance

Reply:
https://github.com/apache/airflow/pull/10949/commits/e4f915efb2b1bd4165d33425799f3bbffd88caa3